### ISSUE TITLE
[NativeAOT LLVM] Using together with .NET 8 SDK

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtimelab#2235]
Troubleshooting of recent issue uncover fact that .NET 8 SDK is not supported.
So my current proposal is advocate following way of consuming NativeAOT-LLVM
```xml
<ItemGroup Condition="$(RuntimeIdentifier) == 'browser-wasm'">
  <!-- Remove existing ILCompilers -->
  <KnownILCompilerPack Remove="Microsoft.DotNet.ILCompiler" />
  <!-- Add NativeAOT LLVM. -->
  <KnownILCompilerPack Include="Microsoft.DotNet.ILCompiler.LLVM"
    TargetFramework="net7.0"
    ILCompilerPackNamePattern="runtime.**RID**.Microsoft.DotNet.ILCompiler.LLVM"
    ILCompilerPackVersion="8.0.0"
    ILCompilerRuntimeIdentifiers="browser-wasm;win-x64" />

     <!-- This line needed only for wildcard consumption -->
    <PackageReference Include="Microsoft.DotNet.ILCompiler.LLVM" Version="8.0.0-*" />
</ItemGroup>
```

Now a bit about how that's working.
Detection of supported ILC package happens here by matching TFM  with KnownILCompilerPack
https://github.com/dotnet/sdk/blob/d8b549b85299046654ba8e9bd0caff0eed507ae6/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L613-L618
and then added as implicit package references here
https://github.com/dotnet/sdk/blob/d8b549b85299046654ba8e9bd0caff0eed507ae6/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L690-L697

I initially [thought](https://github.com/dotnet/sdk/pull/31398) that first place can be extended to perform ILC package selection based on the TFM + RID, but second place insert additioal reference on ILCompiler which causing issues.

So this led me to following solution where I remove existing KnownILCompilerPack and create new one in suited for NativeAOT-LLVM.

If  https://github.com/dotnet/sdk/pull/31398 lands then removal of existing ILCompiler does not needed.

I did not dig deeper why we cannot use NativeAOT-LLVM without PublishAot, maybe that would be more viable solution with less divergence from main.

**Comment by SingleAccretion** [comment]
Thank you for looking into this! I'll share my thinking on the problem space.

At a high level, this can be approached from 2 angles:
1) Add the ability to publish using the LLVM package into SDK - write logic such that careful shimming is possible.
2) Decouple LLVM-based publishing from the SDK. This would still require changes to the SDK logic, unfortunately, if we want to support `PublishAot == true` (which we do), but they would be in a form of "add a switch to disable warnings/errors". I believe we can make it work for `PublishAot == false` without SDK changes, by reverting to the old package-based scheme of acquiring runtime packs.

I have a feeling `1` may not work well long term as upstream is likely to bake more assumptions into the SDK. `2` is not without its risks either, however. I do not have a great handle on which one of these is ultimately better.

One way to look at this is that it is categorically not possible to avoid some kind of project file manipulation beyond the package references (either shimming of `KnownILCompilerPack`, or setting some hypothetical `PublishAotButIWouldLikeTheSdkToThinkNot` property) because of restore, since that runs `ProcessFrameworkReferences.cs` with the hardcoded pack set. From this it follows that `1` and `2` both have equally suboptimal user experience, so it's just about the maintenance.

There is also https://github.com/dotnet/sdk/pull/28495 which would help us a lot if finished.

**Comment by jkotas** [comment]
> we can make it work for PublishAot == false without SDK changes, by reverting to the old package-based scheme of acquiring runtime packs.

I think  PublishAot==false + old package based scheme would work best for as-long-as the project lives in runtimelab.

**Comment by yowl** [comment]
@kant2002 you might want to try this again with the latest packages.

**Comment by kant2002** [comment]
Don't worry, I keep close eye on this :)

**Comment by kant2002** [comment]
.NET 7.0.200 working
.NET 8 preview 2 don't with error `error MSB4057: The target "PrepareForILLink" does not exist in the project`

```
C:\nativeaotwasm\hellojs\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-alpha.1.23208.2\build\Microsoft.NETCore.Native.targets(226,7): error MSB4057: The target "PrepareForILLink" does not exist in the project. [C:\nativeaotwasm\hellojs\hellojs.csproj]
```

**Comment by yowl** [comment]
Thanks, yes agree, does removing the line ` BeforeTargets="_RunILLink"` line 228, maybe, in `Microsoft.NetCore.Native.targets` fix it?

**Comment by kant2002** [comment]
That does not working. For .NET 8 (because it rely on `ProcessFrameworkReferences` task and `KnownILLinkPack` I have to add `<PackageReference Include="Microsoft.NET.ILLink.Tasks" Version="8.0.0-*" />`.

Now I have to use 
```
<ItemGroup>
  <PackageReference Include="Microsoft.DotNet.ILCompiler.LLVM" Version="8.0.0-*" />
  <PackageReference Include="runtime.win-x64.Microsoft.DotNet.ILCompiler.LLVM" Version="8.0.0-*" />
  <PackageReference Include="Microsoft.NET.ILLink.Tasks" Version="8.0.0-*" />
</ItemGroup>
```

But that combination does not work with .NET 7 SDK

**Comment by kant2002** [comment]
Okay. I seems to have working workaround.
```
<ItemGroup>
  <PackageReference Include="Microsoft.DotNet.ILCompiler.LLVM" Version="8.0.0-*" />
  <PackageReference Include="runtime.win-x64.Microsoft.DotNet.ILCompiler.LLVM" Version="8.0.0-*" />
  <PackageReference Include="Microsoft.NET.ILLink.Tasks" Version="8.0.0-*" Condition="'$(ILLinkTargetsPath)' == ''" />
</ItemGroup>
```

**Comment by RReverser** [comment]
This simplifies so much compared to the previous template 😍

**Comment by yowl** [comment]
> > we can make it work for PublishAot == false without SDK changes, by reverting to the old package-based scheme of acquiring runtime packs.
> 
> I think PublishAot==false + old package based scheme would work best for as-long-as the project lives in runtimelab.

I should be able to use `%(PackageDefinitions.Name)` ?

**Comment by hez2010** [comment]
You can simply pass /p:PublishTrimmed=true as well to fix this.

**Comment by RReverser** [comment]
> Okay. I seems to have working workaround.

So this worked for me when you published it but seems it got broken over the last week or so:

```
C:\Users\me\Documents\NativeAOTWasm\helloworld\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-alpha.1.23215.1\build\
Microsoft.NETCore.Native.Publish.targets(73,5): error : Add a PackageReference for 'runtime.browser-x64.Microsoft.DotN
et.ILCompiler.LLVM' to allow cross-compilation for wasm [C:\Users\me\Documents\NativeAOTWasm\helloworld\helloworld.csp
roj]
```

I guess I need to find a pinned version to revert to, as adding said package reference does not help (it doesn't resolve).

**Comment by RReverser** [comment]
Looks like 8.0.0-alpha.1.23208.2 is the last one that works w/o modifications, going to pin to that.

**Comment by hez2010** [comment]
@RReverser You can set IlcHostPackagePath in property manually to workaround this issue, e.g. `<IlcHostPackagePath>C:\Users\<username>\.nuget\packages\runtime.win-x64.microsoft.dotnet.ilcompiler.llvm\<version></IlcHostPackagePath>`

**Comment by RReverser** [comment]
Yeah I don't want to hardcode paths in a shared repo like that, I'm okay with pinned version for now.

**Comment by SteveSandersonMS** [comment]
To avoid hardcoding a path I worked around it like this:

```xml
	<Target Name="FixIlcHostPackageName" BeforeTargets="ImportRuntimeIlcPackageTarget">
		<PropertyGroup>
			<IlcHostPackageName>runtime.win-x64.Microsoft.DotNet.ILCompiler.LLVM</IlcHostPackageName>
		</PropertyGroup>
	</Target>
```

**Comment by yowl** [comment]
With the latest packages this should be fixed.  Add
```
<PropertyGroup>
  <PublishTrimmed>true</PublishTrimmed>
</PropertyGroup>
```
to your `csproj`

**Comment by NCLnclNCL** [comment]
> <PublishTrimmed>true</PublishTrimmed>

with net8 preview 5, you dont need it

**Comment by RReverser** [comment]
Seeing some new build issue on `8.0.0-*`, but only with `-c Release`, debug mode works fine:

```
  Generating native code
  /usr/bin/sh: 2: /tmp/MSBuildTemprreverser/tmp3710694853c443faab955f051a8dd6ad.exec.cmd: \tools\ilc: not found
/home/rreverser/wasiconsole/.packages/microsoft.dotnet.ilcompiler.llvm/8.0.0-preview.7.23414.1/build/Microsoft.NETCore.Native.targets(324,5): error MSB3073: The command ""\tools\\ilc" @"obj/Release/net8.0/browser-wasm/native/wasiconsole.ilc.rsp"" exited with code 127. [/home/rreverser/wasiconsole/wasiconsole.csproj]
```

Although it's been couple of months since I touched NativeAOT, maybe this is not new and just me not having something configured locally?

**Comment by SingleAccretion** [comment]
Looks a lot like the problem that was fixed in #2334.

**Comment by RReverser** [comment]
Hm that PR seems reasonably old so the fix should be shipped in the latest version that I'm using?

**Comment by RReverser** [comment]
The error itself in the PR description looks different too.

**Comment by SingleAccretion** [comment]
Agree it actually looks different. It is also odd that it'd depend on the configuration. Could you share a binlog?

**Comment by RReverser** [comment]
Actually nevermind, I think I changed something unrelated - it does fail on Debug too.

**Comment by SingleAccretion** [comment]
One thing that I should have noted sooner - we never had a Unix build of the host compiler working, and this looks suspiciously like a Unix system.

**Comment by RReverser** [comment]
Oh. Yes, it is.

> we never had a Unix build of the host compiler working

Weird, I thought I did build NativeAOT Wasm experiments successfully on Unix in the past?.. Unless I always did it outside WSL, but that would be weird as I still have some projects in WSL configured for NativeAOT...

**Comment by SingleAccretion** [comment]
> Weird, I thought I did build NativeAOT Wasm experiments successfully on Unix in the past?.. Unless I always did it outside WSL, but that would be weird as I still have some projects in WSL configured for NativeAOT...

What I believe was "working" in the past is host build of the compiler targeting not-WASM (i. e. basically the same compiler that's in the mainline `PublishAot` experience). There are a few issues related to the work required (e. g. #1890, #2119, ...).

**Comment by RReverser** [comment]
Hm I only experimented with Wasm... but anyway, at least now I know what the issue is, thanks. That's a bit unfortunate.

**Comment by yowl** [comment]
Unix did work at one point, but it was many years ago.  When we started moving the compilation over to RyuJIT it broke and hasn't worked since then.

Edit, A bit more info, I was never able to get the LLVM c/c++ headers to play nicely with the Unix build of clrjit.  Things may have got better since then.

**Comment by RReverser** [comment]
Ok I'm compiling for Windows again for now, but seeing a different error that I don't remember (with [RReverser/minisample](https://github.com/RReverser/minisample)):

```
MSBuild version 17.7.1+971bf70db for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  minisample -> C:\Users\me\Documents\minisample\bin\Release\net7.0\browser-wasm\minisample.dll
  Generating native code
  LLVM compilation to IR finished in 0.74 seconds
  LLVM generation of bitcode finished in 1.48 seconds
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  16 errors generated.
emcc : error : 'C:/Users/me/Documents/emsdk/upstream/bin\clang++.exe -target wasm32-unknown-emscripten -fvisibility=default -m
llvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -DEMSCRIPTEN --sysroot=C:\Users\me\Doc 
uments\emsdk\upstream\emscripten\cache\sysroot -Xclang -iwithsysroot/include\fakesdl -Xclang -iwithsysroot/include\compat -c -
fwasm-exceptions -g3 -O3 obj\Release\net7.0\browser-wasm\native\minisample.1.bc -o obj\Release\net7.0\browser-wasm\native\mini 
sample.1.o' failed (returned 1) [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-preview.7.2 
3417.1\build\ExecWrapper.proj]
C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-preview.7.23417.1\build\ExecWrapper.proj(3,5
): error MSB3073: The command ""C:/Users/me/Documents/emsdk/upstream/emscripten/emcc.bat" "obj\Release\net7.0\browser-wasm\nat 
ive\minisample.1.bc" -o "obj\Release\net7.0\browser-wasm\native\minisample.1.o" -c -fwasm-exceptions -g3 -s USE_SDL=0 -O3" exi
ted with code 1.
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr
eview.7.23417.1\build\ExecWrapper.proj]
  <unknown>:0: error: symbol '' can not be placed in a different section
EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
eview.7.23417.1\build\ExecWrapper.proj]
  4 errors generated.
[...more of the same...]
```

Is this because I'm building a library instead of an executable perhaps? Or because I'm using latest emcc (3.1.44)?

**Comment by SingleAccretion** [comment]
> Or because I'm using latest emcc (3.1.44)?

Most likely this (#2376, #2379).

**Comment by RReverser** [comment]
Thanks. I'll try to downgrade. I wonder if the packages (e.g. `Microsoft.DotNet.ILCompiler.LLVM`) should just download its own copy of Emscripten, sort of like `Wasi.Sdk` package does for native WASI SDK. That would prevent any version conflicts.

**Comment by yowl** [comment]
Yeah, I got the same error trying to upgrade emscripten.  Haven't looked into the cause yet.

**Comment by RReverser** [comment]
> I wonder if the packages (e.g. `Microsoft.DotNet.ILCompiler.LLVM`) should just download its own copy of Emscripten, sort of like `Wasi.Sdk` package does for native WASI SDK. That would prevent any version conflicts.

It would be particularly convenient because emsdk doesn't allow having multiple versions of Emscripten installed, so if you want to use latest version on some projects and pinned one on others (like NativeAOT), you _have_ to either re-download a different version on each switch, or have a separate copy in different folder. Adds a bit of a headache to manage it manually.

**Comment by SingleAccretion** [comment]
> Carrying our own EMSDK

Such would be ideal; it is just believed to be non-trivial effort.

One low-cost improvement for the user experience could be to add an `emcc --version` check.

**Comment by RReverser** [comment]
> Such would be ideal; it is just believed to be non-trivial effort.

Hm assuming it's similar to what Wasi.Sdk does, this doesn't seem too bad / too complicated? https://github.com/dotnet/dotnet-wasi-sdk/blob/2dbb00c779180873d3ed985e59e431f56404d8da/src/Wasi.Sdk/build/Wasi.Sdk.targets#L245-L262

**Comment by SingleAccretion** [comment]
Perhaps? I have not personally investigated it; I was/am/should be assuming that we'd need to package the SDK into a NuGet, like the upstream workloads do (indeed, it would be ideal if we could reuse an upstream package). I am not sure how we feel about downloading things "manually".

**Comment by RReverser** [comment]
Ah yeah that would be a bit more complicated. I was only thinking of downloading emsdk / emscripten from official Github releases.

**Comment by NCLnclNCL** [comment]
> Ok I'm compiling for Windows again for now, but seeing a different error that I don't remember (with [RReverser/minisample](https://github.com/RReverser/minisample)):
> 
> ```
> MSBuild version 17.7.1+971bf70db for .NET
>   Determining projects to restore...
>   All projects are up-to-date for restore.
>   minisample -> C:\Users\me\Documents\minisample\bin\Release\net7.0\browser-wasm\minisample.dll
>   Generating native code
>   LLVM compilation to IR finished in 0.74 seconds
>   LLVM generation of bitcode finished in 1.48 seconds
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   16 errors generated.
> emcc : error : 'C:/Users/me/Documents/emsdk/upstream/bin\clang++.exe -target wasm32-unknown-emscripten -fvisibility=default -m
> llvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -DEMSCRIPTEN --sysroot=C:\Users\me\Doc 
> uments\emsdk\upstream\emscripten\cache\sysroot -Xclang -iwithsysroot/include\fakesdl -Xclang -iwithsysroot/include\compat -c -
> fwasm-exceptions -g3 -O3 obj\Release\net7.0\browser-wasm\native\minisample.1.bc -o obj\Release\net7.0\browser-wasm\native\mini 
> sample.1.o' failed (returned 1) [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-preview.7.2 
> 3417.1\build\ExecWrapper.proj]
> C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-preview.7.23417.1\build\ExecWrapper.proj(3,5
> ): error MSB3073: The command ""C:/Users/me/Documents/emsdk/upstream/emscripten/emcc.bat" "obj\Release\net7.0\browser-wasm\nat 
> ive\minisample.1.bc" -o "obj\Release\net7.0\browser-wasm\native\minisample.1.o" -c -fwasm-exceptions -g3 -s USE_SDL=0 -O3" exi
> ted with code 1.
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr
> eview.7.23417.1\build\ExecWrapper.proj]
>   <unknown>:0: error: symbol '' can not be placed in a different section
> EXEC : error : cannot compile inline asm [C:\Users\me\Documents\minisample\.packages\microsoft.dotnet.ilcompiler.llvm\8.0.0-pr 
> eview.7.23417.1\build\ExecWrapper.proj]
>   4 errors generated.
> [...more of the same...]
> ```
> 
> Is this because I'm building a library instead of an executable perhaps? Or because I'm using latest emcc (3.1.44)?

How to use it instead of mscv

**Comment by RReverser** [comment]
Ok so one more thing I wanted to play with is trying to use the new `wasi-wasm` target to build a library with custom imports/exports. It appears that, unlike in browser-wasm, `LibraryImport` for imports & `UnmanagedCallersOnly` for exports currently don't have any effect and wasi-wasm target always produces a WASI executable (with `_start` export and only standard WASI imports) instead.

Is this a known limitation / should I raise a separate issue for this?

**Comment by SingleAccretion** [comment]
`DllImport`s should definitely work so that is definitely not know. `UnmanagedCallersOnly` with `EntryPoint` is known to not produce exports because the linker exports file is not generated/passed.

**Comment by yowl** [comment]
For ` [UnmanagedCallersOnly(EntryPoint = "test:numbers/test#roundtrip-s8")]` you will want to add to your csproj
```
      <CustomLinkerArg Include="-Wl,--export,test:numbers/test!roundtrip-s8" />
```
In an `ItemGroup`

**Comment by RReverser** [comment]
>  because the linker exports file is not generated/passed

Ah good to know it's just a linker issue. Indeed, I just tried and manual workaround via

```xml
<EmccExtraArgs>-Wl,--export=hello</EmccExtraArgs>
```

works for now.

> `DllImport`s should definitely work so that is definitely not know.

Yeah tried both `LibraryImport` and `DllImport` and neither produces a Wasm import on wasi-wasm.

**Comment by RReverser** [comment]
> For ` [UnmanagedCallersOnly(EntryPoint = "test:numbers/test#roundtrip-s8")]` you will want to add to your csproj

Heh thanks, yeah figured out the same for now. I have quite a list of exports, but, luckily, all of them are static.

Imports, however, I can't find workaround for.

**Comment by yowl** [comment]
For the imports add something like
```
        <WasmImport Include="test:numbers/test!roundtrip-s8" />
```

**Comment by yowl** [comment]
Which would import:
```cs
[DllImport("*", EntryPoint = "roundtrip-s8")]
internal static extern int RoundtripS8(int p0);
```

**Comment by RReverser** [comment]
Ah thanks, that worked! I assume this is going to be integrated later? I don't think I needed to specify them manually in browser-wasm?

**Comment by SingleAccretion** [comment]
Yeah, definitely is a bug.

Edit: I think I know why as well: `--unresolved-symbols=ignore-all`.

**Comment by RReverser** [comment]
> Edit: I think I know why as well: `--unresolved-symbols=ignore-all`.

Heh yeah that would do it.

**Comment by RReverser** [comment]
Ok I'm closer at least. Need to figure out what to do next.

Sample C#:

```cs
using System;
using System.Runtime.InteropServices;

public static partial class Lib {
    [DllImport("mylib")]
    public static extern int MyMethod();

	[UnmanagedCallersOnly(EntryPoint = "hello")]
	public static void Hello() {
		Console.WriteLine("Hello from C#");
		MyMethod();
	}
}
```

Sample host using Node:

```js
import { readFile } from 'node:fs/promises';
import { WASI } from 'wasi';
import { argv, env } from 'node:process';

const wasi = new WASI({
  version: 'preview1',
  args: argv.slice(2),
  env,
});

const wasm = await WebAssembly.compile(
  await readFile(new URL('./bin/Release/net7.0/wasi-wasm/publish/minisample.wasm', import.meta.url)),
);
const instance = await WebAssembly.instantiate(wasm, {
  ...wasi.getImportObject(),
  mylib: {
    MyMethod: () => console.log('Hello from JS!'),
  }
});

console.log('Starting WASI...');
wasi.start(instance);
console.log('Invoking Hello');
instance.exports.hello();
```

Looks like the "library" exits immediately after WASI start:

```
> node sample.mjs
(node:2924) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Starting WASI...
```

If I try to comment out WASI start, then instead getting an error that it's not started (presumably because runtime is not initialized):

```
> node sample.mjs
(node:19840) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Invoking Hello
wasm://wasm/015360ca:1


Error: wasi.start() has not been called
    at __wasi_clock_time_get (wasm://wasm/015360ca:wasm-function[10360]:0x21e368)
    at __clock_gettime (wasm://wasm/015360ca:wasm-function[10346]:0x21de76)
    at GCToOSInterface::QueryPerformanceCounter() (wasm://wasm/015360ca:wasm-function[263]:0xbcc4)
    at WKS::gc_heap::init_gc_heap(int) (wasm://wasm/015360ca:wasm-function[477]:0x1cd0e)
    at WKS::GCHeap::Initialize() (wasm://wasm/015360ca:wasm-function[610]:0x45662)
    at RedhawkGCInterface::InitializeSubsystems() (wasm://wasm/015360ca:wasm-function[731]:0x490e6)
    at RhInitialize (wasm://wasm/015360ca:wasm-function[182]:0xabf9)
    at InitializeRuntime() (wasm://wasm/015360ca:wasm-function[144]:0x9485)
    at Thread::ReversePInvokeAttachOrTrapThread(ReversePInvokeFrame*) (wasm://wasm/015360ca:wasm-function[166]:0x9824)
    at RhpReversePInvokeAttachOrTrapThread2 (wasm://wasm/015360ca:wasm-function[171]:0x993f) {
  code: 'ERR_WASI_NOT_STARTED'
}
```

**Comment by SingleAccretion** [comment]
It's almost as if node itself silently crashes somehow (what's the exit code?).

**Comment by RReverser** [comment]
Aha, at least could workaround via `returnOnExit` option that prevents Wasm from being able to actually exit process via `proc_exit`:

```js
const wasi = new WASI({
  version: 'preview1',
  args: argv.slice(2),
  env,
  returnOnExit: true,
});
```

```
> node sample.mjs
(node:20404) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Starting WASI...
Invoking Hello
Hello from C#
Hello from JS!
```

It looks like generated `_start` function in WASI is actually trying to exit the process once it finished executing. That makes sense for `Exe` output, but not for library mode.

**Comment by RReverser** [comment]
> (what's the exit code?).

Ah right, it looks like WASI is exiting with code 71. Can reproduce with or without `returnOnExit: true`:

```
> node sample.mjs   
(node:21988) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Starting WASI...
> echo $LASTEXITCODE  
71
```

with:

```js
console.log('Starting WASI...');
console.log('Exit code:', wasi.start(instance));
console.log('Invoking Hello');
instance.exports.hello();
```

```
> node sample.mjs
(node:21544) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Starting WASI...
Exit code: 71
Invoking Hello
Hello from C#
Hello from JS!
```

**Comment by RReverser** [comment]
Any idea what that exit code stands for? Does generated WASI expect any specific preopens / env vars / etc exposed to it maybe?

**Comment by RReverser** [comment]
Aha, [apparently](https://github.com/bytecodealliance/wasmtime/issues/783) 71 is EX_OSERR:

```c
#define EX_OSERR        71      /* system error (e.g., can't fork) */
```

**Comment by yowl** [comment]
Might just be some "random" value on the stack.

**Comment by RReverser** [comment]
Seems a bit too specific match to be random. Quoting from linked thread:

> In this case it's a process exit status, so it's not an errno value. 71 is sometimes EX_OSERR.
> 
> It might be wasm's startup code, which uses EX_OSERR in the init code to report unlikely failures like __wasi_environ_get etc. failing, or malloc failing before main. It's possible that if your environment doesn't have much memory, the first call to memory.grow prompted by the first malloc call could fail. If so, updating to a new wasi-libc may help, as we avoid calling memory.grow until we actually need it.

**Comment by RReverser** [comment]
Decided to try and dump WASI calls it makes:

```
Starting WASI...
WASI invocation { prop: 'environ_sizes_get', args: [ 881820, 881816 ] }
WASI invocation { prop: 'environ_get', args: [ 887200, 881840 ] }
WASI invocation { prop: 'fd_prestat_get', args: [ 3, 881816 ] }
WASI invocation { prop: 'fd_prestat_dir_name', args: [ 3, 887520, 2 ] }
WASI invocation { prop: 'fd_prestat_get', args: [ 4, 881816 ] }
WASI invocation { prop: 'args_sizes_get', args: [ 881816, 881820 ] }
WASI invocation { prop: 'args_get', args: [ 887600, 887520 ] }
WASI invocation { prop: 'proc_exit', args: [ 71 ] }
Exit code: 71
```

So it's clearly an explicit `proc_exit`, somewhere after reading args, but I'm not familiar enough with code to know where to look / how to find out why it's passing this exit code.

**Comment by SingleAccretion** [comment]
It looks like there are two problems:
1) `/p:NativeLib=Shared` isn't producing "a reactor". Edit: found that we're missing `-mexec-model=reactor`.
2) This lib-not-lib seems to fail before it hits the `unreachable` that the main function (missing due to `NativeLib=Shared`) was turned into.

**Comment by RReverser** [comment]
> * `/p:NativeLib=Shared` isn't producing "a reactor".

Out of curiosity thought I'd try to add `-mexec-model=reactor` to EmccExtraArgs but that doesn't seem to change anything.

**Comment by SingleAccretion** [comment]
> `EmccExtraArgs`

Is Browser/Emscripten-only. There is a `LinkerArg` item group that works across both targets.

Adding:
```xml
<ItemGroup>
  <LinkerArg Include="-mexec-model=reactor" />
</ItemGroup>
```
Did produce `.wasm` that looked more like a reactor on disassembly (it had `_initialize`, for example) for me.

**Comment by RReverser** [comment]
> Out of curiosity thought I'd try to add `-mexec-model=reactor` to EmccExtraArgs but that doesn't seem to change anything.

I actually looked at the wrong wasm file, sorry. Looks like `EmccExtraArgs` works regardless of target. But I suppose `LinkerArg` is a bit cleaner in terms of meaning.

**Comment by RReverser** [comment]
Ok I changed JS code to use it as a reactor (`wasi.initialize` instead of `wasi.start`), and now it initializes without issues but other exports now fail instead:

```
Starting WASI...
WASI invocation { prop: 'environ_sizes_get', args: [ 881948, 881944 ] }
WASI invocation { prop: 'fd_prestat_get', args: [ 3, 881944 ] }
Exit code: undefined
Invoking Hello
wasm://wasm/0153e7c6:1


RuntimeError: unreachable
    at _initialize (wasm://wasm/0153e7c6:wasm-function[27]:0x5bcc)
    at InitializeRuntime() (wasm://wasm/0153e7c6:wasm-function[143]:0x943c)
    at Thread::ReversePInvokeAttachOrTrapThread(ReversePInvokeFrame*) (wasm://wasm/0153e7c6:wasm-function[165]:0x97e7)
    at RhpReversePInvokeAttachOrTrapThread2 (wasm://wasm/0153e7c6:wasm-function[170]:0x9902)
    at RhpReversePInvoke (wasm://wasm/0153e7c6:wasm-function[171]:0x9979)
    at minisample_Lib__Hello (wasm://wasm/0153e7c6:wasm-function[3179]:0xea273)
    at file:///C:/Users/me/Documents/minisample/sample.mjs:31:18
```

Looks like other exports also try to invoke `_initialize` even though it was previously called - maybe something there doesn't have a guard to prevent it from being invoked twice and that's why it traps?

**Comment by SingleAccretion** [comment]
It looks like a consequence of #2364, which was a workaround for cases where `_initialize` does not get invoked. Can you avoid invoking it manually?

**Comment by RReverser** [comment]
> Can you avoid invoking it manually?

With couple of hacks I could but it wouldn't be standard usage, as Node.js expects to be able to call `_initialize` during its own initialisation, then retrieve memory and such from the instance, and only then you should be able to call exports.

**Comment by SingleAccretion** [comment]
Right. I think we could make a better workaround here.

**Comment by RReverser** [comment]
By hacks I meant something like this:

```js
console.log('Starting WASI...');
console.log('Exit code:', wasi.initialize({
  exports: {
    ...instance.exports,
    _initialize: () => console.log('Preventing _initialize'),
  }
}));
console.log('Invoking Hello');
instance.exports.hello();
```

This does work but not ideal yeah.

```
Starting WASI...
Preventing _initialize
Exit code: undefined
Invoking Hello
WASI invocation { prop: 'environ_sizes_get', args: [ 881516, 881512 ] }
WASI invocation { prop: 'fd_prestat_get', args: [ 3, 881512 ] }
WASI invocation { prop: 'clock_time_get', args: [ 1, 1n, 881400 ] }
WASI invocation { prop: 'clock_time_get', args: [ 1, 1n, 881480 ] }
WASI invocation { prop: 'fd_write', args: [ 1, 880440, 1, 880436 ] }
Hello from C#
Hello from JS!
```

**Comment by RReverser** [comment]
Extracted couple of discussed issues into separate GH issues for easier tracking.

**Comment by RReverser** [comment]
One more thing: I just opened same project in a new console and was surprised it stopped compiling with some obscure errors:

```
C:\Program Files\dotnet\sdk\8.0.100-preview.7.23376.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.
targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\Use 
rs\me\Documents\minisample\minisample.csproj]
  minisample -> C:\Users\me\Documents\minisample\bin\Release\net7.0\wasi-wasm\minisample.dll
clang : error : no such file or directory: 'obj/Release/net7.0/wasi-wasm/native/minisample.0.o' [C:\Users\me\Documents\minisam
ple\minisample.csproj]
clang : error : no such file or directory: 'obj/Release/net7.0/wasi-wasm/native/minisample.1.o' [C:\Users\me\Documents\minisam
ple\minisample.csproj]
clang : error : no such file or directory: 'obj/Release/net7.0/wasi-wasm/native/minisample.2.o' [C:\Users\me\Documents\minisam 
ple\minisample.csproj]
clang : error : no such file or directory: 'obj/Release/net7.0/wasi-wasm/native/minisample.3.o' [C:\Users\me\Documents\minisam 
ple\minisample.csproj]
clang : error : no such file or directory: 'obj/Release/net7.0/wasi-wasm/native/minisample.4.o' [C:\Users\me\Documents\minisam 
ple\minisample.csproj]
```

And the directory it looks in has a bunch of `.bc` files but not `.o` files.

Eventually I realised I don't have emsdk loaded in the environment, so as soon as I activated `emsdk_env`, it worked.

But then, I don't understand - why does it need Emscripten for wasi-wasm target if it already has access to and uses WASI SDK Clang for actual linking? Couldn't it use same Clang for bc -> o compilation as well? Or are there significant differences between the two and only the former can handle the bitcode produces by NativeAOT-LLVM?

**Comment by SingleAccretion** [comment]
It's because of a mysterious OOM issue.

See https://github.com/WebAssembly/wasi-sdk/issues/326.

**Comment by RReverser** [comment]
That looks fun 😬

I see the example there uses wasm32-wasi-threads, which is a relatively new target. I'm curious - did the issue occur with the regular wasm32-wasi as well? (just to potentially isolate it)

**Comment by SingleAccretion** [comment]
I don't think we know.

**Comment by RReverser** [comment]
I just tried switching it locally, but getting same error as above. Is it possible to improve error message for now? I wonder why it doesn't fail due to missing `WasmCompilerPath` executable before it gets to linking.

**Comment by SingleAccretion** [comment]
> I just tried switching it locally, but getting same error as above. Is it possible to improve error message for now?

Definitely, we'll do that together with the other fixes.

**Comment by RReverser** [comment]
Also purely out of curiosity, did you try Emscripten's STANDALONE_WASM or PURE_WASI mode instead of Wasi SDK? I know it's not as complete - they have some methods not yet implemented in terms of pure WASI - but if it does work, I wonder if it would be easier to manage a single toolchain instead of two.

Additionally, it would allow to leverage Emscripten's support of Wasm EH already even on WASI target, without waiting for Wasi SDK to implement it as well.

**Comment by SingleAccretion** [comment]
> Also purely out of curiosity, did you try Emscripten's STANDALONE_WASM or PURE_WASI mode instead of Wasi SDK? I know it's not as complete - they have some methods not yet implemented in terms of pure WASI - but if it does work, I wonder if it would be easier to manage a single toolchain instead of two.

AFAIK the answer is no, although actually only @yowl knows. There is a benefit to having just one toolchain, however, upstream also uses WASI SDK, so it may be a drawback too.

> Additionally, it would allow to leverage Emscripten's support of Wasm EH already even on WASI target, without waiting for Wasi SDK to implement it as well.

Note that we don't need C++ exception support to implement managed exception, only actual WASM EH instructions as the ABI we use is fully custom (and evolving with #2374...). Native runtime code also doesn't use exception handling (and is usually built with EH support turned off), so we are well-positioned portability-wise in this aspect.

**Comment by RReverser** [comment]
> Note that we don't need C++ exception support to implement managed exception, only actual WASM EH instructions as the ABI we use is fully custom (and evolving with #2374...). Native runtime code also doesn't use exception handling (and is usually built with EH support turned off), so we are well-positioned portability-wise in this aspect.

Ah nice, good to know it's going to be fully independent. It's just that current output has Emscripten's EH imports, so I assume it's still using C++ - in which case `-fwasm-exceptions` via Emscripten could help even for `wasi-wasm` target.

**Comment by SingleAccretion** [comment]
> I assume it's still using C++

It's a little hack to get it to both compile and run EH-free code (actually, `throw`-free code)... Since `-fwasm-exceptions` would cause validation failures in Wasm-EH-less engines.

The near-term plan for EH is to implement it via emulation (as you have once suggested could be done), and ever so gradually transition to Wasm EH as engines begin to implement it.

**Comment by RReverser** [comment]
Potentially seeing one more issue, not sure yet - might be missing something. Looks like `[DllImport]` in a dependency (if they're not in the top-level assembly) are ignored. I tried putting `<WasmImport ...>` in their `.csproj`, in top-level `.csproj`, but nothing changes it so far - imports just don't appear in the generated wasm.

It works fine when it's in the top-level / main project, just not in a dependency.

**Comment by RReverser** [comment]
Same for exports actually, looks like functions marked with `[UnmanagedCallersOnly]` in a dependency library don't "survive" till linking (`-Wl,--export...` at top level has no effect).

**Comment by SingleAccretion** [comment]
> It works fine when it's in the top-level / main project, just not in a dependency.

This is expected, see [#83396](https://github.com/dotnet/runtime/pull/83396). The short version is this needs another `ItemGroup`:
```xml
<ItemGroup>
  <UnmanagedEntryPointsAssembly Include="DependencyAssembly" />
</ItemGroup>
```
I will add a note to the usage document.

> Looks like [DllImport] in a dependency (if they're not in the top-level assembly) are ignored

This is quite odd. Were those imports rooted properly?

**Comment by RReverser** [comment]
> This is expected, see [#83396](https://github.com/dotnet/runtime/pull/83396). The short version is this needs another `ItemGroup`:

Ohh thanks. I tried to google solutions but didn't find that issue.

> Were those imports rooted properly?

I did include them via `<Import Project />` and then tried to just move actual `<WasmImport>`s into the top-level csproj if that's what you mean, but no difference.

**Comment by RReverser** [comment]
Oh, the `UnmanagedEntryPointsAssembly` helped with both exports and imports as well. I suspect imports were just optimised out as they weren't referenced by any exports - because there were no exports before the fix.

**Comment by SingleAccretion** [comment]
> I suspect imports were just optimised out as they weren't referenced by any exports - because there were no exports before the fix.

Right - that's what I meant by "rooting".

**Comment by RReverser** [comment]
Ah fair enough. Yeah so exports were the root (heh no pun intended) of the problem.

While I have you, is there a way to correctly include extra sources or object files in a NuGet package? For now I'm just doing what I did with Wasi.Sdk - including `bindings.c` in the package and passing it to the linker with:

```xml
<LinkerArg Include="$(MSBuildThisFileDirectory.Replace('\', '/'))../bindings.c" />
```

But this relies on the fact that linker happens to be a compiler as well, which is not ideal, and it does mean that each project using this assembly will duplicate the compilation too. Not a big deal, but I wonder if it's possible to somehow invoke just the compilation to object file in a NuGet package? I tried using `<Exec />` with `$(WasmCompilerPath)` but looks like the latter variable is not exposed.

I also tried just doing `clang -target wasm32-wasi -c bindings.c -o bindings.o` in `<Exec />` and that works but I'd rather make sure I use the same compiler and flags NativeAOT-LLVM is using for other sources, and not just any Clang that happens to be on the path.

**Comment by SingleAccretion** [comment]
If you already have a static library / object file then it is possible to use it via `NativeLibrary`, if that is what you're asking:
```xml
<ItemGroup>
  <NativeLibrary Include="<path>/bindings.o" />
</ItemGroup>
```
Of course, static libraries are rather brittle due to version mismatch issues.

Edit: if the question is what the right approach is for compiling to such object files in the first place, there is no good/supported answer right now. What you've detailed are some things that could/may/may not work. It's a problem similar to the distribution of regular native binaries which you compile from source on platforms without uniform toolchains.

**Comment by RReverser** [comment]
> It's a problem similar to the distribution of regular native binaries which you compile from source on platforms without uniform toolchains.

Right, but I think one difference and benefit of Wasm is that it's the same binary / object file on any platform. As long as it was compiled against same libc at least. So in theory it should be possible for the compiler to produce object files (which are just relocatable Wasm) & compile them & ship as part of arbitrary package, not just top-level app.

**Comment by SingleAccretion** [comment]
> Right, but I think one difference and benefit of Wasm is that it's the same binary / object file on any platform. As long as it was compiled against same libc at least

I think they need to use the same version of the toolchain as well (or one "appropriately close"). I do not actually know at what level is the backwards compatibility requirement enforced - certainly, you would expect the linker to not stop accepting old object files, but LLVM IR (and thus anything LTO) will of course be continually broken. Emscripten I believe expects exact matches.

More practically, regarding the compiler flags, I don't believe we require anything specific from _foreign code_ at this point in time.

Edit: well - we don't support `-shared`/"PIE" code, but that will become evident quickly.

**Comment by RReverser** [comment]
LLVM IR does change, yeah. I was only thinking of non-LTO code which is good enough for simple cases when you need to drop out to C (e.g. to export variable from Wasm, which I believe is currently not possible from C#, or to override Wasi imports with noop functions). But you're right that if we want LTO, it's better to ship the source.

I guess for now I'll use `<NativeLibrary Include="(path)/bindings.c" />` which I can ship in the assembly, but maybe there should be at least a separate `ItemGroup` item for C files to be passed to the compiler so that we don't rely on linker being able to accept C code as well.

**Comment by RReverser** [comment]
Alright, it finally all comes together. Very experimental and not something we can ship yet, obviously - both due to nature of the project and because we'll need that emulated exception support before we can test it on something more complex - but finally ported one of canonical examples on our project from `Wasi.Sdk` to NativeAOT-LLVM to the point that it compiles & I can actually invoke necessary exports.

First invocation went down 2000x and subsequent invocations went down 150x compared to Wasi.Sdk's interpreted version 😍 Kudos to the team for the amazing work on this project! Will continue keeping an eye on further developments.

**Comment by RReverser** [comment]
> Unix did work at one point, but it was many years ago. When we started moving the compilation over to RyuJIT it broke and hasn't worked since then.
> 
> Edit, A bit more info, I was never able to get the LLVM c/c++ headers to play nicely with the Unix build of clrjit. Things may have got better since then.

I thought I'd give this a try just to see which issues arise, since this is the next blocker.

I had almost no problems building LLVM and then `clr.wasmjit+clr.aot+libs`, aside from one missing `#include`.

Architecture-specific `./build.sh -arch wasm -os wasi  -s libs -c Debug` part also worked fine.

First problem I'm stuck on for now arises when trying to build `clr.aot` as well, as in `./build.sh -arch wasm -os wasi  -s clr.aot+libs -c Debug` like in the docs:

```
...
  Repo successfully built.
  Product binaries are available at /home/rreverser/runtimelab/artifacts/bin/coreclr/wasi.wasm.Debug
  Microsoft.Interop.SourceGeneration -> /home/rreverser/runtimelab/artifacts/bin/Microsoft.Interop.SourceGeneration/Debug/netstandard2.0/Microsoft.Interop.SourceGeneration.dll
  System.Private.CoreLib.Generators -> /home/rreverser/runtimelab/artifacts/bin/System.Private.CoreLib.Generators/Debug/netstandard2.0/System.Private.CoreLib.Generators.dll
  LibraryImportGenerator -> /home/rreverser/runtimelab/artifacts/bin/LibraryImportGenerator/Debug/netstandard2.0/Microsoft.Interop.LibraryImportGenerator.dll
  System.Private.CoreLib -> /home/rreverser/runtimelab/artifacts/bin/System.Private.CoreLib/ref/Debug/net8.0/System.Private.CoreLib.dll
CSC : error CS2001: Source file '/home/rreverser/runtimelab/artifacts/obj/coreclr/wasi.wasm.Debug/nativeaot/Runtime/Portable/AsmOffsetsPortable.cs' could not be found. [/home/rreverser/runtimelab/src/coreclr/nativeaot/System.Private.CoreLib/src/System.Private.CoreLib.csproj]

Build FAILED.

CSC : error CS2001: Source file '/home/rreverser/runtimelab/artifacts/obj/coreclr/wasi.wasm.Debug/nativeaot/Runtime/Portable/AsmOffsetsPortable.cs' could not be found. [/home/rreverser/runtimelab/src/coreclr/nativeaot/System.Private.CoreLib/src/System.Private.CoreLib.csproj]
    0 Warning(s)
    1 Error(s)
```

This doesn't seem LLVM-specific error, but rather some build config shenanigans. I see that file should be generated by `Runtime/Portable/CMakeLists.txt`, but it seems as if that file is never invoked / included by build.

The condition to include it seems to be here

https://github.com/dotnet/runtimelab/blob/ed15503ca4e65ae4087ff282bf186d13da469d14/src/coreclr/nativeaot/Runtime/CMakeLists.txt#L313-L318

and it depends only on `-a wasm`, so I don't see why it would be included correctly on Windows but not on Unix. In the artifacts folder I can also see that `artifacts/obj/coreclr/wasi.wasm.Debug/nativeaot/Runtime/Full` is created but `/Portable` is not, which further confirms that it's like that condition is not triggered correctly for some reason.

**Comment by yowl** [comment]
`shenanigans` Right, this is a super annoying file which doesn't always seem to get generated depending on what has been built and/or failed previously.  First thing to try is to clean, then `build clr.aot+clr.wasmjit+libs ...`

**Comment by RReverser** [comment]
> First thing to try is to clean, then build clr.aot+clr.wasmjit+libs ...

Hm does `clr.wasmjit` need to be built for `-arch wasm -os wasi` as well? Looking at the docs it seemed it only needs to be built for the actual host, while `clr.aot+libs` for both host and Wasm.

**Comment by SingleAccretion** [comment]
The build output makes it seem that the native runtime pieces are not built at all, which would explain the missing `AsmOffsets.cs`.

If I were debugging this, I would diff `build-runtime.cmd` in this branch against upstream to see what was needed to enable the wasm build.

**Comment by RReverser** [comment]
> The build output makes it seem that the native runtime pieces are not built at all

FWIW if you're referring to `...` in the beginning, I just trimmed that output before posting for convenience. But I'll recheck after trying @yowl's suggestion.

**Comment by yowl** [comment]
> > First thing to try is to clean, then build clr.aot+clr.wasmjit+libs ...
> 
> Hm does `clr.wasmjit` need to be built for `-arch wasm -os wasi` as well? Looking at the docs it seemed it only needs to be built for the actual host, while `clr.aot+libs` for both host and Wasm.

Sorry, missed it was the `wasm` build that was failing.  No you don't want the `clr.wasmjit` there, but if you had a previous failure for some missing include, then I wouldn't be surprised if a clean "fixed" it.

**Comment by RReverser** [comment]
> but if you had a previous failure for some missing include, then I wouldn't be surprised if a clean "fixed" it

The missing include is in `UnixContext.h`: it refers to `ucontext_t` but doesn't `#include <ucontext.h>` that declares that type. I think that one was a legitimate issue and adding the include fixed it. I probably should make a PR for this one.

Also now that I looked upstream, I see it has the same line I had to add: https://github.com/dotnet/runtime/blob/ce3b14da2fcc4c97ebdfbbd26d47a9e441db5297/src/coreclr/nativeaot/Runtime/unix/UnixContext.h#L7

> No you don't want the `clr.wasmjit` there

Ah cool. So to double-check, now I've cleaned the artifacts and just doing `./build.sh -arch wasm -os wasi  -s clr.aot+libs -c Debug`, this time skipping the non-Wasm build I did in the previous attempt.

[...later...] 

Ok yeah now it fails with a lot of C errors instead. I guess it picked up some artifacts from non-Wasm build before? The docs were a bit unclear on the order of Wasm and non-Wasm builds. Anyway, at least this is not build config issue anymore - thanks - so will look into those C errors now.

**Comment by SingleAccretion** [comment]
> The docs were a bit unclear on the order of Wasm and non-Wasm builds

Nominally, it should not matter and the two builds should be completely independent (we are building artifacts for two very different platforms, after all). Theory != practice, of course.

**Comment by yowl** [comment]
I'm not sure that it is picking up artifacts from the other build, as @SingleAccretion says, should be independent and in my experience they are.  I just suspect (and never investigated to be fair), that something is not quite right with the dependencies and it skips the step that includes the generation of this file because some other artifact that did work is present.

**Comment by RReverser** [comment]
> I'm not sure that it is picking up artifacts from the other build,

Out of curiosity I just tried again and I'm getting same error if I try to build `./build.sh clr.wasmjit+clr.aot+libs -c Debug` and then try `./build.sh -arch wasm -os wasi  -s clr.aot+clr.wasmjit+libs -c Debug` again.

If I'm only doing `./build.sh -arch wasm -os wasi  -s clr.aot+clr.wasmjit+libs -c Debug` on a clean build, I'm never running into the AsmOffsetsPortable issue, so I'm still leaning into the theory that builds somehow conflict with each other. I might be wrong, but it would at least explain the symptoms.

**Comment by RReverser** [comment]
> If I'm only doing `./build.sh -arch wasm -os wasi -s clr.aot+clr.wasmjit+libs -c Debug` on a clean build, I'm never running into

or maybe not "never"... 😅

**Comment by yowl** [comment]
Is there a reason why you have `clr.wasmjit` in the `-arch wasm -os wasi` build, you don't need or want that?

**Comment by RReverser** [comment]
Sorry, this time it was my turn, I copied the wrong thing into the issue. Indeed, I'm not building `clr.wasmjit` in Wasm build.

**Comment by RReverser** [comment]
...and now I'm consistently getting the `AsmOffsetsPortable.cs` error again even with clean build followed by `./build.sh -s clr.aot+libs -c Debug -a wasm -os wasi`. Even though that's what seemed to have fixed the issue for me above.

I love build systems.

**Comment by RReverser** [comment]
> If I were debugging this, I would diff `build-runtime.cmd` in this branch against upstream to see what was needed to enable the wasm build.

I'll try this next.

**Comment by RReverser** [comment]
Ok at least `./build.sh -s clr.aot -c Debug -a wasm -os wasi --ninja` is now compiling too. ~~Next up, libs.~~ `clr.aot+libs` too.

**Comment by RReverser** [comment]
Looking at wasmjit config:

https://github.com/dotnet/runtimelab/blob/9ca9fc84d29a8aa68ad4434ff920c5ee60be9f7b/src/coreclr/jit/CMakeLists.txt#L719-L720

Do I understand right from those lines that it's only built for the browser and not WASI? That probably doesn't make sense unless `wasmjit` is optional somehow?

**Comment by SingleAccretion** [comment]
So, the `OS` moniker here refers to the target OS. Some Jit's are built with hardcoded target OSes, others use dynamic checks. The `wasmjit` is of the latter variety and should really be called `clrjit_universal_wasm32.dll`, it's just a historical artifact that it is not now.

**Comment by RReverser** [comment]
Ah ok thanks. I'll try to get `clr.wasmjit` to build now. It kinda did earlier, but it was only because `CLR_CMAKE_BUILD_LLVM_JIT` was not actually set by the script so it wasn't really built yet unfortunately.

**Comment by RReverser** [comment]
Right now I'm mostly struggling with all the missing headers in PAL. I admit I don't really understand this system or why it (in pal/inc/rt/cpp) duplicates STL headers but with dummy contents in some places and custom replacements in others.

There also seems to be some `USE_STL` build variable that sounds like it should switch to C++ STL but it still tries to and fails to find many headers in that PAL folder instead :/

**Comment by RReverser** [comment]
And even if I add those dummy headers, then I'm getting errors like `use of undeclared identifier 'size_t'`, `unknown type name 'uint32_t'`, etc, which suggests they shouldn't be dummy? Why doesn't build system use real C / C++ headers instead like it does on Windows?

**Comment by SingleAccretion** [comment]
> I admit I don't really understand this system

Neither do I, truth be told. I know there have been (a number of) conversations in the past about all these problems and what the ways to solve them are, but I do not recall where.

@jkotas would you have advice on the direction here?

**Comment by yowl** [comment]
https://github.com/dotnet/runtimelab/issues/1797 is the old issue.  https://github.com/dotnet/runtimelab/compare/feature/NativeAOT-LLVM...yowl:llvm-linux#diff-af66f6b851332d1e155eef9bf358246005bd77648b156261a74ef36ee90ea1e9R237 are a couple of the changes I made around STL.  All a bit old now.

**Comment by RReverser** [comment]
@yowl Oh wow, the size of that changeset. I guess I was optimistic 😅

I have the shell script changes already which look similar and enabled me to build runtime packages; I don't have the `if (CLR_CMAKE_BUILD_LLVM_JIT)` parts though - I'll try and see how far that gets me without all the rest.

**Comment by jkotas** [comment]
> @jkotas would you have advice on the direction here?

CoreCLR PAL for non-Windows platforms emulates Windows SDK behaviors that are incompatible with standard headers on Linux. It means that a single .cpp file cannot include both PAL headers and Linux standard headers at the same time.

@AaronRobinsonMSFT and others have been working towards cleaning this up and removing this limitation.

**Comment by RReverser** [comment]
> PAL used to build CoreCLR for non-Windows platforms emulates Windows SDK behaviors that are incompatible with standard headers on Linux.

I think I saw that mentioned about emulating Windows SDK and that makes sense for platform-specific APIs, but why would C++ STL itself be so different that it needs to replace/emulate those headers as well? Is this down to differences between MSVC and GCC/Clang headers?

**Comment by jkotas** [comment]
Yes. Also, CoreCLR PAL does not emulate everything. It only emulates parts needed by CoreCLR. C++ STL depends on much more.

**Comment by RReverser** [comment]
> Yes. Also, CoreCLR PAL does not emulate everything. It only emulates parts needed by CoreCLR. C++ STL depends on much more.

Right, and that mismatch seems to be what leads to complications. When you say it's being worked on you mean the code will be compatible with any STL or that more things will be emulated over time?

**Comment by jkotas** [comment]
The goal is to remove all PAL constructors that conflict with standard headers, so that it is possible to use both the PAL and standard headers at the same time.

**Comment by RReverser** [comment]
I got only couple more compilation errors fixed but I think I'm going to give up for now :( The interaction between PAL and STL in the build system is beyond my comprehension at the moment, too many weird errors.

**Comment by jkotas** [comment]
> The interaction between PAL and STL in the build system is beyond my comprehension at the moment, too many weird errors.

Yes, this is something that needs to be cleaned up upstream first. It does not make sense to do this in this repo/branch.
